From 653377582ce34a3595a7758a98916c611c97cf81 Mon Sep 17 00:00:00 2001
From: Youff <asa.jazal@gmail.com>
Date: Mon, 6 Apr 2026 00:15:50 +0700
Subject: [PATCH 1/8] cpuidle: menu: Simplify latency constraint handling via unified helper

Refactor the menu governor selection mechanism to retrieve CPU latency
requirements using the unified cpuidle_governor_latency_req() helper. 
This replaces manual evaluations of PM QoS and device-specific raw resume 
latencies, streamlining the logic and allowing the removal of redundant 
header inclusions.
---
 drivers/cpuidle/governors/menu.c | 10 +---------
 1 file changed, 1 insertion(+), 9 deletions(-)

diff --git a/drivers/cpuidle/governors/menu.c b/drivers/cpuidle/governors/menu.c
index c7820a09fa65..436807afcd83 100644
--- a/drivers/cpuidle/governors/menu.c
+++ b/drivers/cpuidle/governors/menu.c
@@ -12,7 +12,6 @@
 
 #include <linux/kernel.h>
 #include <linux/cpuidle.h>
-#include <linux/pm_qos.h>
 #include <linux/time.h>
 #include <linux/ktime.h>
 #include <linux/hrtimer.h>
@@ -21,7 +20,6 @@
 #include <linux/sched/loadavg.h>
 #include <linux/sched/stat.h>
 #include <linux/math64.h>
-#include <linux/cpu.h>
 
 /*
  * Please note when changing the tuning values:
@@ -298,15 +296,13 @@ static int menu_select(struct cpuidle_driver *drv, struct cpuidle_device *dev,
 		       bool *stop_tick)
 {
 	struct menu_device *data = this_cpu_ptr(&menu_devices);
-	struct device *device = get_cpu_device(dev->cpu);
-	int latency_req = pm_qos_request(PM_QOS_CPU_DMA_LATENCY);
+	int latency_req = cpuidle_governor_latency_req(dev->cpu);
 	int i;
 	int first_idx;
 	int idx;
 	unsigned int interactivity_req;
 	unsigned int expected_interval;
 	unsigned long nr_iowaiters, cpu_load;
-	int resume_latency = dev_pm_qos_raw_read_value(device);
 	ktime_t delta_next;
 
 	if (data->needs_update) {
@@ -314,10 +310,6 @@ static int menu_select(struct cpuidle_driver *drv, struct cpuidle_device *dev,
 		data->needs_update = 0;
 	}
 
-	/* resume_latency is 0 means no restriction */
-	if (resume_latency && resume_latency < latency_req)
-		latency_req = resume_latency;
-
 	/* Special case when user has set very strict latency requirement */
 	if (unlikely(latency_req == 0)) {
 		*stop_tick = false;

From 2d96740b10b59ca3d388b034f4b76f10fd3848ee Mon Sep 17 00:00:00 2001
From: Youff <asa.jazal@gmail.com>
Date: Mon, 6 Apr 2026 00:22:28 +0700
Subject: [PATCH 2/8] rosemary_defconfig: Enable TEO and standard cpuidle infrastructure

Update the defconfig to enable generic CPU idle subsystem support,
allowing multiple drivers. Build both the standard Menu governor and 
the Timer Events Oriented (TEO) governor to enhance CPU idle precision 
and power efficiency options.
---
 arch/arm64/configs/rosemary_defconfig | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/arch/arm64/configs/rosemary_defconfig b/arch/arm64/configs/rosemary_defconfig
index a85289a4b5d2..5866f12c11b4 100644
--- a/arch/arm64/configs/rosemary_defconfig
+++ b/arch/arm64/configs/rosemary_defconfig
@@ -26,6 +26,14 @@ CONFIG_MTK_SCHED_TRACERS=n
 CONFIG_MTK_SCHED_CPU_PREFER=n
 CONFIG_MTK_SCHED_BIG_TASK_MIGRATE=n
 CONFIG_SCHED_CASS=y
+CONFIG_CPU_IDLE=y
+CONFIG_CPU_IDLE_MULTIPLE_DRIVERS=y
+# CONFIG_CPU_IDLE_GOV_LADDER is not set
+CONFIG_CPU_IDLE_GOV_MENU=y
+# CONFIG_CPU_IDLE_GOV_MTK is not set
+CONFIG_CPU_IDLE_GOV_MTK_MENU=y
+# CONFIG_CPU_IDLE_NO_MTK is not set
+CONFIG_CPU_IDLE_GOV_TEO=y
 # CONFIG_SLUB_DEBUG is not set
 CONFIG_MACH_MT6785=y
 CONFIG_CFQ_GROUP_IOSCHED=y

From d50637ae3c0eabd0e56281c5e6056c956985007d Mon Sep 17 00:00:00 2001
From: Youff <asa.jazal@gmail.com>
Date: Mon, 6 Apr 2026 00:27:34 +0700
Subject: [PATCH 3/8] cpuidle: teo: Elevate governor rating to override defaults

Increase the rating of the TEO governor from 19 to 150. This adjustment
ensures the kernel prioritizes and automatically selects TEO over other 
built-in governors like Menu when initializing CPU idle states.
---
 drivers/cpuidle/governors/teo.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/cpuidle/governors/teo.c b/drivers/cpuidle/governors/teo.c
index f376f4caccb4..69eb812ac5a5 100644
--- a/drivers/cpuidle/governors/teo.c
+++ b/drivers/cpuidle/governors/teo.c
@@ -524,7 +524,7 @@ static int teo_enable_device(struct cpuidle_driver *drv,
 
 static struct cpuidle_governor teo_governor = {
 	.name =		"teo",
-	.rating =	19,
+	.rating =	150,
 	.enable =	teo_enable_device,
 	.select =	teo_select,
 	.reflect =	teo_reflect,

From 69ffe82aa823fedce0dd0b03078279f00d42f9cb Mon Sep 17 00:00:00 2001
From: Youff <asa.jazal@gmail.com>
Date: Mon, 6 Apr 2026 00:40:12 +0700
Subject: [PATCH 4/8] cpuidle: teo: Add structural documentation line inside governor block

Include an inline documentation note within the teo_enable_device
function header block to track upcoming structural adjustments.
---
 drivers/cpuidle/governors/teo.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/cpuidle/governors/teo.c b/drivers/cpuidle/governors/teo.c
index 69eb812ac5a5..0746c17ce70a 100644
--- a/drivers/cpuidle/governors/teo.c
+++ b/drivers/cpuidle/governors/teo.c
@@ -507,6 +507,7 @@ static void teo_reflect(struct cpuidle_device *dev, int state)
  * teo_enable_device - Initialize the governor's data for the target CPU.
  * @drv: cpuidle driver (not used).
  * @dev: Target CPU.
+ * update commit title
  */
 static int teo_enable_device(struct cpuidle_driver *drv,
 			     struct cpuidle_device *dev)

From af1c6e3e81e7b0f02480d2e23556cc2f9c3ecae3 Mon Sep 17 00:00:00 2001
From: Youff <asa.jazal@gmail.com>
Date: Mon, 6 Apr 2026 01:49:32 +0700
Subject: [PATCH 5/8] rosemary_defconfig: Isolate cpuidle implementation from MTK extensions

Enforce upstream cpuidle behavior by enabling CONFIG_CPU_IDLE_NO_MTK.
This removes the proprietary MediaTek menu governor additions from the build
to avoid side effects and rely fully on TEO.
---
 arch/arm64/configs/rosemary_defconfig | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/arch/arm64/configs/rosemary_defconfig b/arch/arm64/configs/rosemary_defconfig
index 5866f12c11b4..d4fcca218031 100644
--- a/arch/arm64/configs/rosemary_defconfig
+++ b/arch/arm64/configs/rosemary_defconfig
@@ -30,9 +30,7 @@ CONFIG_CPU_IDLE=y
 CONFIG_CPU_IDLE_MULTIPLE_DRIVERS=y
 # CONFIG_CPU_IDLE_GOV_LADDER is not set
 CONFIG_CPU_IDLE_GOV_MENU=y
-# CONFIG_CPU_IDLE_GOV_MTK is not set
-CONFIG_CPU_IDLE_GOV_MTK_MENU=y
-# CONFIG_CPU_IDLE_NO_MTK is not set
+CONFIG_CPU_IDLE_NO_MTK=y
 CONFIG_CPU_IDLE_GOV_TEO=y
 # CONFIG_SLUB_DEBUG is not set
 CONFIG_MACH_MT6785=y

From e245bc383e653ea2136aa07a8153fb45d9b33926 Mon Sep 17 00:00:00 2001
From: Youff <asa.jazal@gmail.com>
Date: Mon, 6 Apr 2026 01:59:43 +0700
Subject: [PATCH 6/8] cpuidle: Kconfig: Set CPU_IDLE_NO_MTK to true by default

Modify the cpuidle Kconfig structural rules to set CPU_IDLE_NO_MTK 
to default to yes. This prevents the default inclusion of proprietary 
MediaTek driver logic across generic platform targets.
---
 drivers/cpuidle/Kconfig | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/cpuidle/Kconfig b/drivers/cpuidle/Kconfig
index b7a3fc519f62..177cfffbb527 100644
--- a/drivers/cpuidle/Kconfig
+++ b/drivers/cpuidle/Kconfig
@@ -51,12 +51,14 @@ config CPU_IDLE_GOV_MTK_MENU
 
 config CPU_IDLE_NO_MTK
 	bool "Disable MTK related cpuidle driver / governor"
+	default y
 	help
 	  MTK added many proprietary cpuidle drivers and governors.
 	  And some of them set to be default for all projects.
 	  But not all (MTK) projects use these driver.
 	  Set this option to avoid using MTK driver / gorvernor by default.
 
+
 config DT_IDLE_STATES
 	bool
 

From a94dfcd02fdef0f1494e1f6d7ad404d7232795db Mon Sep 17 00:00:00 2001
From: Youff <asa.jazal@gmail.com>
Date: Mon, 6 Apr 2026 02:02:52 +0700
Subject: [PATCH 7/8] Revert "cpuidle: Kconfig: Set CPU_IDLE_NO_MTK to true by default"

This reverts commit e245bc383e653ea2136aa07a8153fb45d9b33926 to avoid global 
Kconfig dependency overrides, restoring the option's original behavior.
---
 drivers/cpuidle/Kconfig | 2 --
 1 file changed, 2 deletions(-)

diff --git a/drivers/cpuidle/Kconfig b/drivers/cpuidle/Kconfig
index 177cfffbb527..b7a3fc519f62 100644
--- a/drivers/cpuidle/Kconfig
+++ b/drivers/cpuidle/Kconfig
@@ -51,14 +51,12 @@ config CPU_IDLE_GOV_MTK_MENU
 
 config CPU_IDLE_NO_MTK
 	bool "Disable MTK related cpuidle driver / governor"
-	default y
 	help
 	  MTK added many proprietary cpuidle drivers and governors.
 	  And some of them set to be default for all projects.
 	  But not all (MTK) projects use these driver.
 	  Set this option to avoid using MTK driver / gorvernor by default.
 
-
 config DT_IDLE_STATES
 	bool
 

From b5b05f44147963e88996aa66a18dbed3d32bba2f Mon Sep 17 00:00:00 2001
From: Youff <asa.jazal@gmail.com>
Date: Mon, 6 Apr 2026 02:05:23 +0700
Subject: [PATCH 8/8] Revert "rosemary_defconfig: Isolate cpuidle implementation from MTK extensions"

This reverts commit af1c6e3e81e7b0f02480d2e23556cc2f9c3ecae3 to restore 
MediaTek's custom governor entries inside the target defconfig structure.
---
 arch/arm64/configs/rosemary_defconfig | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/configs/rosemary_defconfig b/arch/arm64/configs/rosemary_defconfig
index d4fcca218031..5866f12c11b4 100644
--- a/arch/arm64/configs/rosemary_defconfig
+++ b/arch/arm64/configs/rosemary_defconfig
@@ -30,7 +30,9 @@ CONFIG_CPU_IDLE=y
 CONFIG_CPU_IDLE_MULTIPLE_DRIVERS=y
 # CONFIG_CPU_IDLE_GOV_LADDER is not set
 CONFIG_CPU_IDLE_GOV_MENU=y
-CONFIG_CPU_IDLE_NO_MTK=y
+# CONFIG_CPU_IDLE_GOV_MTK is not set
+CONFIG_CPU_IDLE_GOV_MTK_MENU=y
+# CONFIG_CPU_IDLE_NO_MTK is not set
 CONFIG_CPU_IDLE_GOV_TEO=y
 # CONFIG_SLUB_DEBUG is not set
 CONFIG_MACH_MT6785=y
 
